### ISSUE TITLE
Demo video reproduction

### ISSUE THREAD

**Issue by sillsill777** [erizmr/SPH_Taichi#22]
Hi, Thanks for your lovely project.
Anyway I've runned `python run_simulation.py --scene_file ./data/scenes/dragon_bath.json`  and what I saw was

![capture](https://user-images.githubusercontent.com/37329078/217728444-2a749d7f-e539-4c32-ac66-99e85e8b1d9b.png)

I was expecting like the video you've uploaded(with realistic water and objects).
Is it somewhat related to the `GLFW Error 65537: The GLFW library is not initialized`? Which I got while running the code.

Thanks.

**Comment by erizmr** [comment]
Hi @sillsill777 ,
Thanks for your interest in this project.
The image you posted here is the expected behavior. The GUI embedded in the simulator is only for fast preview use.
For higher quality rendering, postprocessing is required. The video uploaded in the repo is rendered using Houdini Mantra with the data output by the simulator. 

You can set `exportPly` in the  `dragon_bath.json` to `true` so that the simulator can output the particle data in `.ply` files. Then load this files using the software you like for rendering.

**Comment by erizmr** [comment]
I have updated the script a bit. Could you please try again with the latest master? Thanks

**Comment by sillsill777** [comment]
![cap](https://user-images.githubusercontent.com/37329078/217826539-f5d13870-d803-4ebd-8671-6face32cdd9f.png)

It seems that `ps.object_collection[r_body_id]` does not have key corresponding to `mesh`.
 `ps.object_collection[r_body_id]` has the keys presented below.


`dict_keys(['objectId', 'geometryFile', 'translation', 'rotationAxis', 'rotationAngle', 'scale', 'velocity', 'density', 'color', 'isDynamic', 'particleNum', 'voxelizedPoints'])`

Sorry to bothering you

**Comment by erizmr** [comment]
No worries. It should work now. Please check the latest master.

**Comment by sillsill777** [comment]
Hi there! I was checking your latest master. I set `exportPly` to True and then at some point scene blows up. I attached gif to describe what happend.

![bad-gif](https://user-images.githubusercontent.com/37329078/217992916-9b5841b0-8168-46b4-8d98-aa4af6060e2c.gif)

Also I tried it several time, file `particle_object_0_******` inside the `dragon_bath_output` always ends at `particle_object_0_000034`. I assume at that  point program blows up?

Anyway if I set `eportPly` to false, program works as a charm. Also I attach gif for the good result when I set `exportPly` to false. 

![ezgif com-video-to-gif](https://user-images.githubusercontent.com/37329078/217993750-93b3ceab-643d-4785-8320-90d5c9783c19.gif)

Thanks again.

**Comment by erizmr** [comment]
Thanks for reporting this. I tried to reproduce it and found that it only happens with the latest Taichi version 1.4.0. It works fine using Taichi 1.3.0. For workaround, you can try with Taichi 1.3.0. I am doing some further investigation to figure out what happens in the new version Taichi.

**Comment by sillsill777** [comment]
Hi there, with your modification, I was able to attain result enven with `exportPly` to true. (taichi version v.1.4.1). 
After I downloaded Houdini and somewhat managed to import ply files. (`file` -> `import` -> `geometry` -> `particle_object_0_$F6.ply (0-426)`) And this is what I get.

![tmp](https://user-images.githubusercontent.com/37329078/218241956-8dafcb0b-f290-4ee3-82db-6813ffb8b0d7.png)

And the problem was with that bunch of buttons, I cannot figure out what I have to do.
I clearly don't want you to walk me through every details to render scene in your demo video.  But not to that extent, can I get any resouces or even some online video lecture link, guiding how to render like your demo video with the `.ply` files? If you can't find such a thing it is totally fine, at this point I already owe you too much.

Thanks again for all.

**Comment by erizmr** [comment]
Hi, there are basically several steps for rendering the demo video:
1. Import the geometry files (e.g., `.ply`) 
2. Restrict the particles inside a bound box to avoid some unexpected edge cases using a bound box, e.g., a `delete` node
3. Perform fluid surface reconstruction using `Particle Fluid Surface` node.
4. Use the Houdini built-in CPU based mantra renderer to render images. It may takes few hours to render the whole sequence. 
Here is a video link for tutorial of fluid surface reconstruction and rendering: https://www.youtube.com/watch?v=AmnKMmEor9I

**Comment by sillsill777** [comment]
Sorry for the super late reply. I was late because I wanted to show you my own result. With your amazing support and devotion I was able to reproduce the result as in your  demo video. This is the final result.

![default_rendered](https://user-images.githubusercontent.com/37329078/222072869-7d693ec4-7b10-4ecc-ad52-d603595eaede.gif)

 Again thanks for your **amazing project** and kindness!! Becase of this, I was able to understand what  SPH is. 
I wish your best luck!

Best regards,
TaeHoon Yoon
